### PR TITLE
Remove caching to enable flexible face value evaluations with variable-functors

### PR DESCRIPTION
**Pull request by grmnptr**
Closes. #21508

### CODE REVIEW

File: modules/navier_stokes/src/variables/INSFVVelocityVariable.C [hunk 1]
```diff
@@ -65,6 +65,41 @@ INSFVVelocityVariable::INSFVVelocityVariable(const InputParameters & params) : I
 }
 
 #ifdef MOOSE_GLOBAL_AD_INDEXING
+
+const ADReal &
+INSFVVelocityVariable::getExtrapolatedBoundaryFaceValue(const FaceInfo & fi) const
+{
+  if (_two_term_boundary_expansion)
+    if (isFullyDevelopedFlowFace(fi))
```
Comment (lindsayad) [inline modules/navier_stokes/src/variables/INSFVVelocityVariable.C:73]: Can just group these two ifs together and reduce indentation

File: modules/navier_stokes/src/variables/INSFVVelocityVariable.C [hunk 2]
```diff
@@ -65,6 +65,41 @@ INSFVVelocityVariable::INSFVVelocityVariable(const InputParameters & params) : I
...
+    }
+
+  return INSFVVariable::getExtrapolatedBoundaryFaceValue(fi);
+}
+
+const VectorValue<ADReal> &
+INSFVVelocityVariable::uncorrectedAdGradSln(const FaceInfo & fi, const bool correct_skewness) const
+{
+  if (_two_term_boundary_expansion)
+    if (isFullyDevelopedFlowFace(fi))
```
Comment (lindsayad) [inline modules/navier_stokes/src/variables/INSFVVelocityVariable.C:92]: same

File: framework/include/utils/FVUtils.h [hunk 3]
```diff
@@ -35,22 +45,14 @@ loopOverElemFaceInfo(const Elem & elem,
   {
     const Elem * const candidate_neighbor = elem.neighbor_ptr(side);
 
-    bool elem_has_info;
+    bool elem_has_info = elemHasFaceInfo(elem, candidate_neighbor);
```
Comment (lindsayad) [inline framework/include/utils/FVUtils.h:48]:
```suggestion
    const bool elem_has_info = elemHasFaceInfo(elem, candidate_neighbor);
```

File: framework/include/utils/FVUtils.h [hunk 4]
```diff
@@ -35,22 +45,14 @@ loopOverElemFaceInfo(const Elem & elem,
...
 
     std::set<const Elem *> neighbors;
 
+    bool inactive_neighbor_detected = candidate_neighbor ? !candidate_neighbor->active() : false;
```
Comment (lindsayad) [inline framework/include/utils/FVUtils.h:52]:
```suggestion
    const bool inactive_neighbor_detected = candidate_neighbor ? !candidate_neighbor->active() : false;
```

File: modules/navier_stokes/src/variables/INSFVVelocityVariable.C [hunk 5]
```diff
@@ -65,6 +65,41 @@ INSFVVelocityVariable::INSFVVelocityVariable(const InputParameters & params) : I
...
+      const Elem * const elem = std::get<0>(tup);
+      const Point vector_to_face = std::get<2>(tup) ? (fi.faceCentroid() - fi.elemCentroid())
+                                                    : (fi.faceCentroid() - fi.neighborCentroid());
+
+      _temp_face_value = uncorrectedAdGradSln(fi) * vector_to_face + getElemValue(elem);
+
+      return _temp_face_value;
+    }
+
+  return INSFVVariable::getExtrapolatedBoundaryFaceValue(fi);
```
Comment (lindsayad) [inline modules/navier_stokes/src/variables/INSFVVelocityVariable.C:85]: Let's then also put this in an `else` branch. I think it's more readable that way. It's nice to reduce indentation, but sometimes that can make the code less clear

File: modules/navier_stokes/src/variables/INSFVVelocityVariable.C [hunk 6]
```diff
@@ -65,6 +65,41 @@ INSFVVelocityVariable::INSFVVelocityVariable(const InputParameters & params) : I
...
+  if (_two_term_boundary_expansion)
+    if (isFullyDevelopedFlowFace(fi))
+    {
+      const auto & cell_gradient = adGradSln(&fi.elem(), correct_skewness);
+      const auto normal = fi.normal();
+      _temp_face_unc_gradient = cell_gradient - (cell_gradient * normal) * normal;
+      return _temp_face_unc_gradient;
+    }
+
+  return INSFVVariable::uncorrectedAdGradSln(fi, correct_skewness);
```
Comment (lindsayad) [inline modules/navier_stokes/src/variables/INSFVVelocityVariable.C:100]: Put in `else` branch

**Review by lindsayad — CHANGES_REQUESTED** [review]
these are excellent changes. Along with them, could we remove the `_temp` data members and instead of return a `const` reference, instead just return a value? The compiler should give us NRVO I hope so shouldn't incur copy construction costs

File: modules/navier_stokes/src/variables/INSFVVelocityVariable.C (hunk 1 above)
Comment (grmnptr) [reply modules/navier_stokes/src/variables/INSFVVelocityVariable.C:73]: The reason why I thought about doing this is that it is easy to check `_two_term_bc==true` but the `FullyFevelopedFlowFace(fi)` requires a warehouse lookup every time it is called. I am not sure if the lookups are cached or how expensive they can be, so this was a micro-optimization from my side. Let me know if you think this has close to zero gain and I'll  merge the two conditions.

Comment (lindsayad) [reply modules/navier_stokes/src/variables/INSFVVelocityVariable.C:73]: You know that c++ short-circuits right? For `if (A && B)`, if `A` is `false`, then `B` is not evaluated. So I don't see any gains here by having two separate ifs

Comment (grmnptr) [reply modules/navier_stokes/src/variables/INSFVVelocityVariable.C:73]: I did not know that, I'll merge the two then. Thanks for the new information!

File: framework/include/utils/FVUtils.h [hunk 7]
```diff
@@ -21,6 +21,16 @@ namespace Moose
 {
 namespace FV
 {
+
+/*
+ * This function infers based on elements if the faceinfo between them
+ * belongs to the element or not.
+ * @param elem Reference to an element
+ * @param neighbor Reference to the neighbor of the element
```
Comment (GiudGiud) [inline framework/include/utils/FVUtils.h:29]:
```suggestion
 * @param neighbor Pointer to the neighbor of the element
```

File: framework/include/utils/GreenGaussGradient.h [hunk 8]
```diff
@@ -235,15 +220,12 @@ greenGaussGradient(const ElemArg & elem_arg,
     mooseAssert(two_term_boundary_expansion,
                 "I believe we should only get singular systems when two-term boundary expansion is "
                 "being used");
-    const auto grad = greenGaussGradient(elem_arg, functor, false, mesh, face_to_value_cache);
+    const auto grad = greenGaussGradient(elem_arg, functor, false, mesh);
 
     // We failed to compute the extrapolated boundary faces with two-term expansion and callers of
     // this method may be relying on those values (e.g. if the caller is
     // getExtrapolatedBoundaryFaceValue) so we populate them here with one-term expansion, e.g. we
     // set the boundary face values to the cell centroid value
```
Comment (GiudGiud) [inline framework/include/utils/GreenGaussGradient.h:228]: remove this

Comment (GiudGiud) [reply framework/include/utils/GreenGaussGradient.h:228]: actually move it instead of remove it I guess

File: framework/src/utils/FVUtils.C [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+#include "MooseVariableFV.h"
+#include "FVUtils.h"
+
+namespace Moose
+{
+namespace FV
+{
+
+bool
```
Comment (GiudGiud) [inline framework/src/utils/FVUtils.C:18]: docstrings

link to the explanation of the rules for FaceInfo in mesh source file

File: framework/src/variables/MooseVariableFV.C [hunk 10]
```diff
@@ -574,24 +562,10 @@ MooseVariableFV<OutputType>::getInternalFaceValue(const FaceInfo & fi,
...
-
-    if (!pr.second)
-      // Insertion didn't happen...we already have a value ready to go
-      return pr.first->second;
-
-    value_pointer = &pr.first->second;
-  }
-  ADReal & value = *value_pointer;
-  value = Moose::FV::linearInterpolation(*this, Moose::FV::makeCDFace(fi, correct_skewness));
+  _temp_face_value =
```
Comment (GiudGiud) [inline framework/src/variables/MooseVariableFV.C:565]: no need for temp_face_value anymore imo

**Review by GiudGiud — CHANGES_REQUESTED** [review]
any noticeable performance hit?

the logic for "two term fails, resort to one term" needs to stay, it's necessary for resilience to users

File: framework/include/utils/FVUtils.h (hunk 7 above)
Comment (grmnptr) [reply framework/include/utils/FVUtils.h]: Fixed!

File: framework/src/utils/FVUtils.C (hunk 9 above)
Comment (grmnptr) [reply framework/src/utils/FVUtils.C:18]: The doxygen now links to the corresponding function in MooseMesh. The docstring is in the h file.

**Comment by grmnptr** [comment]
So if I remove the `tmp_` member variables for the face value and face gradient, we get the runtimes below (using `time command`). The test case is here:
`modules/navier_stokes/test/tests/finite_volume/pins/channel-flow/2d-rc-friction.i`
I refined it to have 153,600 main Dofs and 51,200 aux dofs. 
I averaged the runtimes from 3 consecutive runs. 

current next                                                     | 98.88 s
current changes with keeping the members   | 98.41 s
current changes with removing the members | 99.03 s

**Comment by lindsayad** [comment]
Looks good! Remove away!

File: framework/src/utils/FVUtils.C [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+elemHasFaceInfo(const Elem & elem, const Elem * const neighbor)
```
Comment (GiudGiud) [inline framework/src/utils/FVUtils.C:19]: is there any way you could use this function in MooseMesh when creating the face infos?
It s the same logic

File: framework/src/variables/MooseVariableFV.C [hunk 12]
```diff
@@ -674,44 +629,23 @@ MooseVariableFV<OutputType>::getExtrapolatedBoundaryFaceValue(const FaceInfo & f
...
-    adGradSln(elem);
-
-    it = _face_to_value.find(&fi);
-    mooseAssert(it != _face_to_value.end(),
-                "adGradSln(elem) should have generated the boundary face value for us");
-
-    return it->second;
+    const Point vector_to_face = std::get<2>(tup) ? (fi.faceCentroid() - fi.elemCentroid())
+                                                  : (fi.faceCentroid() - fi.neighborCentroid());
+    boundary_value = adGradSln(elem) * vector_to_face + getElemValue(elem);
```
Comment (GiudGiud) [inline framework/src/variables/MooseVariableFV.C:639]: I think you can reproduce the resilient behavior by returning 0 for the gradient when the two term fails

File: framework/src/variables/MooseVariableFV.C [hunk 13]
```diff
@@ -882,9 +753,9 @@ MooseVariableFV<OutputType>::adGradSln(const FaceInfo & fi, const bool correct_s
   // perform the correction. Note that direction is important here because we have a minus sign.
   // Neighbor has to be neighbor, and elem has to be elem. Hence all the elem_is_elem_one logic
   // above
-  face_grad += ((neighbor_value - elem_value) / fi.dCFMag() - face_grad * fi.eCF()) * fi.eCF();
+  unc_grad += ((neighbor_value - elem_value) / fi.dCFMag() - unc_grad * fi.eCF()) * fi.eCF();
```
Comment (GiudGiud) [inline framework/src/variables/MooseVariableFV.C:756]: but it s corrected now, cant be called unc_grad

**Review by GiudGiud — COMMENTED** [review]
last things then ready in my mind

File: framework/src/variables/MooseVariableFV.C (hunk 12 above)
Comment (grmnptr) [reply framework/src/variables/MooseVariableFV.C:639]: I don't think the gradient is 0 in general if the 2term fails. There might be special cases when that is true but definitely not in general. And we still need/cache the gradient.

Comment (GiudGiud) [reply framework/src/variables/MooseVariableFV.C:639]: yeah you re right. ok good on this.

**Review by GiudGiud — DISMISSED** [review]

**Review by lindsayad — DISMISSED** [review]
My review addressed

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**